### DISCUSSION TITLE
python Wrapper

### DISCUSSION THREAD

**Discussion by clacon20** [su2code/SU2#1503] [Q&A]
Hello!

I've been trying to wrap SU2_DOT in python to interface with a code I've developed. I tried it last week by simply including "SU2_DOT.hpp" in the pySU2.i file, and it was working!! However after having deleted and re-downloaded the code, by simply repeating the same steps (in the same order) I get errors in the compilation, addressing "undefined  references " to functions created in SU2_DOT. 
Can anyone help with this?

Thanks a lot

**Comment by pcarruscag** [comment]
Hello,
This work was started in PR #1300, maybe that already does what you need or maybe @patelha57 and @aa-g can give you some pointers.

**Reply by clacon20** [reply]
Hello,
thanks for the quick reply. I had a look at that conversation but I don't see how that was solver in the end.
By the way my concern is purely on the python wrapping capabilities of SU2. I tried to include SU2_DOT by simply adding the include statement in the pysu2.i interface file and got it working last week, but somehow by simply erasing all binary files and rebuilding the code again it does not seem to be working any more despite repeating exact the same commands. 

By the way I'll try to get in touch with @patelha57 to ask if they managed to do something.

**Reply by clacon20** [reply]
Hello @pcarruscag,

unfortunately I didn't manage to find a solution to my problem. I honestly do not understand why by simply defining a new function to the already wrapped codes (like CDriver.cpp or CSingleZoneDriver.cpp) and recompiling the code, these will not be included in the pysu2 module which is imported in python. 

Is there anything you would suggest? 
Thanks,
Claudio

**Reply by pcarruscag** [reply]
If you are doing it exactly like the other functions I don't see why it would not work, but I've never done too much python wrapper development.
Do you have a branch you can point me to?
Some wild guesses:
Are you defining the functions inline (in the hpp)? Are they free functions or members of those classes? Are they private or public?
(if yes to any, try doing the opposite)

**Reply by clacon20** [reply]
The branch I'm using is the develop (I tried it with the master as well). 
Functions are defined in the header hpp files, and they are member of those classes already wrapped. I made a very simple test with a function printing "hello world" in CDriver (i.e. CDriver::test() {...} ).

I'll check it out if switching from public to private makes a difference.
I really appreciate the quick reply, thanks!

**Reply by pcarruscag** [reply]
try putting the definition in the cpp, do not make them private if they were public.

**Comment by patelha57** [comment]
Some comments regarding PR #1300. Not only did we update the mesh deformation routines, but we also exposed them to Python. We have that same goal in mind: to couple with other Python-based MDAO frameworks, optimization routines, etc. [Here](https://github.com/su2code/SU2/tree/feature_pysu2_DEF/SU2_DEF/src/drivers) are the refactored routines. Some quick remarks:

- **CDeformationDriver** is the forward deformation routine for standard mesh deformation, either for analysis or updating design variables. It supports legacy mesh deformation routines as well as the newer CMeshSolver (depending on how the `DEFORM_MESH` option is set in the SU2 config).

- **CDiscAdjDeformationDriver** is the reverse counterpart, including the projection step that you are interested in. This driver is essentially the SU2_DOT routine.

Once again, these are intended to be used with Python as in your case. We had to add [meson build files](https://github.com/su2code/SU2/blob/feature_pysu2_DEF/SU2_DEF/src/meson.build) and also update [pySU2 files](https://github.com/su2code/SU2/tree/feature_pysu2_DEF/SU2_PY/pySU2) (pySU2.i and pySU2ad.i, and most importantly the [meson.build](https://github.com/su2code/SU2/blob/feature_pysu2_DEF/SU2_PY/pySU2/meson.build) file in that directory) to expose them to Python. I don’t believe it is as simple as what you have already tried and described (i.e. adding include statement into the pySU2.i file) and compiling. Please try with these additional files and let me know if it helps!
 
On a related note, we have another branch called [feature_pysu2_fsi_adjoint](https://github.com/su2code/SU2/tree/feature_pysu2_fsi_adjoint). This is the latest version of our refactoring efforts for pysu2/pysu2ad, and includes a significantly improved [Python API](https://github.com/su2code/SU2/blob/feature_pysu2_fsi_adjoint/SU2_CFD/src/python_wrapper_structure.cpp) with a lot more functionality for analysis and sensitivity analysis. We are in process of starting a new PR for this latest work, so it can be integrated in future releases of SU2.
